core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -1116,62 +708,89 @@ protected boolean shouldExecuteReplication(Settings settings) {
+        public void close() {
+            operationLock.close();
         }
 
-        static IndexShardReferenceImpl createOnReplica(IndexShard indexShard, long primaryTerm) {
-            return new IndexShardReferenceImpl(indexShard, primaryTerm);
+        public boolean isRelocated() {
+            return indexShard.state() == IndexShardState.RELOCATED;
         }
 
         @Override
-        public void close() {
-            operationLock.close();
+        public void failShard(String reason, Throwable e) {
+            try {
+                indexShard.failShard(reason, e);
+            } catch (Throwable suppressed) {
+                e.addSuppressed(suppressed);
+            }
         }
 
         @Override
-        public boolean isRelocated() {
-            return indexShard.state() == IndexShardState.RELOCATED;
+        public Tuple<Response, ReplicaRequest> performOperation(Request request) throws Exception {
+            return shardOperationOnPrimary(request);
         }
 
         @Override
-        public void failShard(String reason, @Nullable Throwable e) {
-            indexShard.failShard(reason, e);
+        public long term() {
+            return indexShard.getPrimaryTerm();
         }
 
         @Override
         public ShardRouting routingEntry() {
             return indexShard.routingEntry();
         }
+    }
+
+    final class ReplicasProxy implements ReplicatedOperation.ReplicasProxy<ReplicaRequest> {
 
         @Override
-        public long opPrimaryTerm() {
-            return indexShard.getPrimaryTerm();
+        public void performOn(ShardRouting replica, ReplicaRequest request, ActionListener<TransportResponse.Empty> listener) {
+            String nodeId = replica.currentNodeId();
+            final DiscoveryNode node = clusterService.state().nodes().get(nodeId);
+            if (node == null) {
+                listener.onFailure(new NoNodeAvailableException("unknown node [" + nodeId + "]"));
+                return;
+            }
+            transportService.sendRequest(node, transportReplicaAction, request, transportOptions,
+                ActionListenerResponseHandler.responseHandler(listener, () -> TransportResponse.Empty.INSTANCE));
+        }
+
+        @Override
+        public void failShard(ShardRouting replica, ShardRouting primary, String message, Throwable throwable,
+                              Runnable onSuccess, Consumer<Throwable> onFailure, Consumer<Throwable> onIgnoredFailure) {
+            shardStateAction.shardFailed(
+                replica, primary, message, throwable,
+                new ShardStateAction.Listener() {
+                    @Override
+                    public void onSuccess() {
+                        onSuccess.run();
+                    }
+
+                    @Override
+                    public void onFailure(Throwable shardFailedError) {
+                        if (shardFailedError instanceof ShardStateAction.NoLongerPrimaryShardException) {
```

Comment:
I'm not too happy with this solution as well. I made it like this so ReplicatedOperation will not have to import stuff ShardStateAction and rely on their semantics. Instead it is made explicit in the it's failShard call backs. From the ReplicatedOperation perspective either there is a proper failure (in which case the operation needs to with NoLongerPrimary, because we failed to do a primary-level operation) or something happened that can safely be ignored. I'm open for better solutions